clang/test/SemaSYCL/free_function_kernel_params_restrictions.cpp

```diff
@@ -0,0 +1,42 @@
+// RUN: %clang_cc1 -internal-isystem %S/Inputs -fsycl-is-device -verify %s
+// This test checks that compiler correctly diagnoses violations of restrictions
+// applied to free function kernel parameters defined by the spec.
+
+#include "sycl.hpp"
+
+class Outer {
+public:
+  class DefinedWithinAClass {
+    int f;
+  };
+};
+
+__attribute__((sycl_device))
+[[__sycl_detail__::add_ir_attributes_function("sycl-single-task-kernel", 0)]]
+void ff_4(Outer::DefinedWithinAClass S1) { // expected-error {{'Outer::DefinedWithinAClass' cannot be used as the type of a kernel parameter}}
```

Comment:
This diagnostic message could be improved; it doesn't explain *why* the type can't be used.

Reply:
I added notes emitted. Hopefully this improves the situation.

Reply:
Yes, this looks good, thank you!